[RFC] Port of the termdebug.vim plugin to neovim terminal feature.

kwon-young opened the pull request:
Port of the termdebug.vim plugin to neovim terminal feature as requested in #7379.
For neovim compatibility, the vim specific calls were replaced with neovim specific calls:
* term_start -> term_open
* term_sendkeys -> jobsend
* term_getline -> getbufline

This is a linux port only of the plugin, because I don't have a windows to test on.
There is still a bug with the Evaluate command.
It always trigger the following error: `E15: Invalid expression: <range>, ''`

I think that before merging this PR, some other feature should make it into neovim:

* [x] #6743 for getting the tty and buffer of a job using nvim_get_chan_info
* [x] merge patch 8.0.1089 for the evaluation feature: #8946. Closed in d0c8dfc
* [x] merge patch 8.0.1090 for \<cexpr\> evaluation: 3d71366af1287f356a812a38f2c0b22230fe409c
* [X] #6619 for having the same floating window evaluation feature when cursor on a variable

![Screenshot_20190515_205609](https://user-images.githubusercontent.com/2647934/57801412-edf8b080-7753-11e9-9076-2cef338ecdb4.png)

janlazo commented:
Can you list all related vim-patches that create/modify `termdebug.vim` and resolve them first?

kwon-young commented:
I've compiled the following list of related vim-patches from the history of termdebug.vim here: https://github.com/vim/vim/commits/master/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim
In reverse chronological order:

* patch 8.0.1725: terminal debugger doesn't handle command arguments
* patch 8.0.1713: terminal debugger doesn't handle arguments 
* patch 8.0.1668: terminal debugger: can't re-open source code window 
* patch 8.0.1655: outdated gdb message in terminal debugger unclear 
* patch 8.0.1599: no error message when gdb does not support debugger 
* patch 8.0.1562: the terminal debugger can't set a breakpoint with the… 
* patch 8.0.1318: terminal balloon only shows one line 
* patch 8.0.1309: cannot use 'balloonexpr' in a terminal 
* patch 8.0.1293: setting a breakpoint in the terminal debugger sometim… 
* patch 8.0.1264: terminal debugger gets stuck in small window 
* patch 8.0.1123: cannot define a toolbar for a window 
* patch 8.0.1107: terminal debugger jumps to non-existing file 
* patch 8.0.1092: terminal debugger can't evaluate expressions 
* patch 8.0.1085: terminal debugger can't set breakpoints 
* patch 8.0.1077: no debugger making use of the terminal window

janlazo commented:
Excluding the `runtime/` changes, are any of those patches N/A to Neovim?

kwon-young commented:
N/A: do you mean Not Applicable ?

Maybe the patch 8.0.1089 for the `range` argument. This modifies some c code in neovim.
I've tried to apply it, but couldn't make it work successfully.

> Vim 8.0 +terminal starts at patch 8.0.0693.

... I suppose, but I don't think this is relevant for this pull request.
The whole point is to use the neovim terminal feature instead of the vim built-in one.

phcerdan commented:
Thanks for working on this @kwon-young :+1:

kwon-young committed:
commit 36257d0f97b396467bef7a5937befd894fb23e31
Author: Kwon-Young Choi <kwon-young.choi@hotmail.fr>
Date:   Sat May 5 16:57:45 2018 +0200

    Port of the termdebug.vim plugin to neovim terminal feature.
    For neovim compatibility,
    The vim specific calls were replaced with neovim specific calls:
      term_start -> term_open
      term_sendkeys -> jobsend
      term_getline -> getbufline
      job_info && term_getjob -> using linux command ps to get the tty

    fix1: forgot to port EndDebug callback to neovim

    fix2: use nvim_get_chan_info to get pty of job
          remove the use of communication buffer by using jobstart instead
          of termopen

    fix3: get gdbbuf using nvim_get_chan_info

mcepl commented:
With [this patch](https://build.opensuse.org/package/view_file/home:mcepl:neovim/neovim/neovim-termdebug-PR8364.patch?expand=1) on the checkout from GitHub master branch, I hoped to get a working termdebug. However, when I run:
```
:packadd! termdebug
:Termdebug
```
I get no error after the first command and the second one raises ``"E492: Not an editor command: Termdebug"``. And yes, the plugin is in ``/usr/share/nvim/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim`` and ``/usr/share/nvim/runtime`` is in ``&rtp``.

kwon-young commented:
I don't really know how packadd works, but in the meantime, you can just download the termdebug.vim file from my branch and do a: `source termdebug.vim` and you should have the `Termdebug` command available.

By the way, I'm waiting for an official release of neovim 4.0 to finish porting the `termdebug` plugin, so I can use floating windows when evaluating variables

mcepl commented:
OK, so yes, there is something wrong with the installation, because ``:packadd termdebug | Termdebug`` doesn't work. When I run ``:source /usr/share/nvim/runtime/pack/dist/opt/termdebug/plugin/termdebug.vim``, then there is ``:Termdebug`` command and it looks like working. Thank you.

mcepl commented:
Right, so I have a solution for myself. For ``:packadd`` the configuration option is ``&packpath`` (not ``&rtp``), so make sure it has correct value (I had ``set packpath=~/.vim`` in my ``~/.config/nvim/init.vim``). With the correct value of the option, ``packadd termdebug|Termdebug`` works.

justinmk commented:
Is there any reason we can't just merge what's here, in the meantime?

mcepl commented:
+1 from me. I haven't made any real testing, but when I start ``:Termdebug`` (with packages from [my openSUSE/Fedora repository](https://build.opensuse.org/package/show/home:mcepl:neovim/neovim)) I get something, which looks correct:
![Snímek z 2019-05-13 13-22-31](https://user-images.githubusercontent.com/198999/57617960-5c861480-7582-11e9-91dc-6f99f31d782f.png)

mcepl commented:
@kwon-young do you plan to push this patch back to vim (with all those ``if has("nvim")``, it feels so)?

kwon-young commented:
> @kwon-young do you plan to push this patch back to vim (with all those `if has("nvim")`, it feels so)?

No, I don't think bram is interested in supporting neovim for termdebug ...

I think the reason I put these `if has("nvim")` was because I wasn't sure if I wanted to make a plugin out of this, to extend its feature set. In that case it would have been a shame to just loose vim support.

> Is there any reason we can't just merge what's here, in the meantime? (just need conflicts to be resolved)

No reason from my part, I would be super excited to make my first contribution to neovim!

Commit:
Merge remote-tracking branch 'upstream/master'

wsdjeg commented:
@kwon-young if this is built-in runtime file of neovim, I think this is no need to check `if has('nvim')`, because this file will not be used in vim. just modified this plugin based on neovim's job api will be better.

mcepl commented:
> @kwon-young if this is built-in runtime file of neovim, I think this is no need to check `if has('nvim')`, because this file will not be used in vim. just modified this plugin based on neovim's job api will be better.

+1 this in the end IS internal neovim implementation. Unless @brammool accepts your compromise (which I don't expect), we will have just to live with two independent implementations, and then KISS principle leads to elimination of those ``if``s.

kwon-young committed:
cleaned up if has('nvim') to remove vim support.
added neovim floating window support for expression evaluation

Commit:
improvred documentation, cleaned up vim menu code, fixed bug when
floating window feature is not available

justinmk commented:
Thank you @kwon-young !

bfredl commented on runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:
```diff
@@ -228,85 +540,352 @@ func s:Evaluate(range, arg)
+
+        execute 'noswapfile edit!' bufname
+
+        setlocal winhl=Normal:CursorLine
```
Not sure we want this,  `NormalFloat` will be automatically used and should be set to something reasonable for hover popups etc.

kwon-young replied:
Okay, I agree that the colourscheme should put a sensible default for NormalFloat.

bfredl commented on runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:
```diff
@@ -228,85 +540,352 @@ func s:Evaluate(range, arg)
+        return
+    endif
+    autocmd! plugin-LC-neovim-close-hover
+    execute winnr . 'wincmd c'
```
call `nvim_win_close(a:win_id)` is more robust, it works even if the user switched tab.

kwon-young replied:
thanks, this makes me think that I should also change the name of that autocmd so that it does not clash with languageclient-neovim...

bfredl commented on runtime/pack/dist/opt/termdebug/plugin/termdebug.vim:
```diff
@@ -228,85 +540,352 @@ func s:Evaluate(range, arg)
+
+        " Give margin
+        let width += 1
+        let lines = [''] + lines + ['']
```
Vertical margin of an entire screen row consumes quite a lot of screen space. Maybe we can remove this for now, and soon we can make this configurable by user with the 'margin' option.

mcepl commented:
Just that this probably needs to be a new pull request. This one has been merged.

bfredl commented:
@mcepl https://github.com/neovim/neovim/pull/10015 is the follow-up.
